In this notebook I will develop some simple code to do some basic cleaning of the raw dataset. Here we want to deal with the issues identified in the EDA notebook, including removing rows where we have missing values (i.e. `?` values) and remove the spaces before the text.

Once we have this I will then put it into a script and use a DVC pipeline to run it.

In [2]:
import pandas as pd

# Load data

We can use the `na_values` argument in the `read_csv` method to add ` ?` as a missing value. This will then replace these with `nan` in the resulting DataFrame. We can also use the `skipinitialspace` argument to ignore the preceeding space before the text. These two options fix most of the issues.

In [24]:
df = pd.read_csv('./../data/census.csv', na_values='?', skipinitialspace=True)

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       30725 non-null  object
 2   fnlgt           32561 non-null  int64 
 3   education       32561 non-null  object
 4   education-num   32561 non-null  int64 
 5   marital-status  32561 non-null  object
 6   occupation      30718 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital-gain    32561 non-null  int64 
 11  capital-loss    32561 non-null  int64 
 12  hours-per-week  32561 non-null  int64 
 13  native-country  31978 non-null  object
 14  salary          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


So we currently still have 32561 samples, but we now see that several features have missing values (e.g. workclass, occupation and native-country). Below we can now see that we have 'nan' values in the native-country feature:

In [20]:
df['native-country'].unique()

array(['United-States', 'Cuba', 'Jamaica', 'India', nan, 'Mexico',
       'South', 'Puerto-Rico', 'Honduras', 'England', 'Canada', 'Germany',
       'Iran', 'Philippines', 'Italy', 'Poland', 'Columbia', 'Cambodia',
       'Thailand', 'Ecuador', 'Laos', 'Taiwan', 'Haiti', 'Portugal',
       'Dominican-Republic', 'El-Salvador', 'France', 'Guatemala',
       'China', 'Japan', 'Yugoslavia', 'Peru',
       'Outlying-US(Guam-USVI-etc)', 'Scotland', 'Trinadad&Tobago',
       'Greece', 'Nicaragua', 'Vietnam', 'Hong', 'Ireland', 'Hungary',
       'Holand-Netherlands'], dtype=object)

# Clean data
## Drop rows with missing values
First we will drop the samples containing missing values. We can do this simply using the `dropna()` DataFrame method.

In [21]:
df.dropna(inplace=True)

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30162 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             30162 non-null  int64 
 1   workclass       30162 non-null  object
 2   fnlgt           30162 non-null  int64 
 3   education       30162 non-null  object
 4   education-num   30162 non-null  int64 
 5   marital-status  30162 non-null  object
 6   occupation      30162 non-null  object
 7   relationship    30162 non-null  object
 8   race            30162 non-null  object
 9   sex             30162 non-null  object
 10  capital-gain    30162 non-null  int64 
 11  capital-loss    30162 non-null  int64 
 12  hours-per-week  30162 non-null  int64 
 13  native-country  30162 non-null  object
 14  salary          30162 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


Now we can see that we only have 30162 samples and none of the features have missing values.

## Remove space before text
We already did this by using the `skipinitialspace` argument when we read in the file. We can check that the column names are also fixed below.

In [26]:
df.columns

Index(['age', 'workclass', 'fnlgt', 'education', 'education-num',
       'marital-status', 'occupation', 'relationship', 'race', 'sex',
       'capital-gain', 'capital-loss', 'hours-per-week', 'native-country',
       'salary'],
      dtype='object')

# Output cleaned file
The only thing left to do is output a new csv file containing the cleaned dataset.

In [27]:
df.to_csv('./../data/census_cleaned_test.csv')

This works fine. We don't actually want to keep this so let's delete the file we just generated.

In [28]:
!rm ./../data/census_cleaned_test.csv